Proofs/Experiments/Max/MaxTandem.lean

```diff
@@ -0,0 +1,898 @@
+  r StateField.PC sn = 0x8c8#64 ∧
+  r StateField.ERR sn = .None ∧
+  sn.program = program ∧
+  sn.mem = s.mem ∧
+  cut sn = false ∧
+  sn.sp = s.sp ∧
+  sn[sn.sp + 28#64, 4] = s[s.sp + 28#64, 4]  ∧
+  sn.x0 = BitVec.zeroExtend 64 (sn[sn.sp + 28#64, 4])  ∧
+  CheckSPAlignment sn :=  by
+  have := program.stepi_eq_0x8c4 h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  simp [pcs]
+
+/--
+info: 'MaxTandem.program.stepi_0x8c4_cut' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8c4_cut
+
+-- 14/15: add  sp, sp, #0x20
+theorem program.stepi_0x8c8_cut (s sn : ArmState)
+  (h_program : s.program = program)
+  (h_pc : r StateField.PC s = 0x8c8#64)
+  (h_err : r StateField.ERR s = StateError.None)
+  (h_sp_aligned : CheckSPAlignment s)
+  (h_step : sn = run 1 s) :
+  r StateField.PC sn = 0x8cc#64 ∧
+  r StateField.ERR sn = .None ∧
+  cut sn = true ∧
+  sn.program = program ∧
+  sn.x0 = s.x0 ∧
+  sn.sp = s.sp + 0x20#64 ∧
+  CheckSPAlignment sn :=  by
+  have := program.stepi_eq_0x8c8 h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  simp (config := {ground := true, decide := true})
+  simp only [fst_AddWithCarry_eq_add, true_and]
+  apply Aligned_BitVecAdd_64_4
+  · assumption
+  · decide
+
+/--
+info: 'MaxTandem.program.stepi_0x8c8_cut' depends on axioms: [propext, Classical.choice, Lean.ofReduceBool, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8c8_cut
+
+end CutTheorems
+
+/--
+rename an old term `old` identifier a new identifier `new`, and replace
+all occurrences of `old` with `new`.
+-/
+macro "rename" old:ident "to" new:ident : tactic =>
+  `(tactic|
+    generalize h_rename : $old = $new <;> subst $old)
+
+/-- define a value `x := val`, and name the hypothesis `heq : x := val` -/
+macro "name" heq:ident ":" x:ident " := " val:term : tactic =>
+  `(tactic| generalize $heq : $val = $x)
+
+
+/-- info: Correctness.cassert {σ : Type} [Sys σ] [Spec' σ] (s0 si : σ) (i : Nat) : Nat × Prop -/
+#guard_msgs in #check Correctness.cassert
+
+/-- info: MaxTandem.cut (s : ArmState) : Bool -/
+#guard_msgs in #check cut
+
+open Lean Elab Meta Tactic in
+/-- If no name is supplied, then clears all inaccessible names.
+If a list of names is supplied, then clear all inaccessible names that have
+*any* one of the strings as prefixes. -/
+elab "clear_named" "[" names:(ident),* "]": tactic =>  do
+  withMainContext do
+    for hyp in (← getLCtx) do
+      -- trace[debug] "name: {hyp.userName}"
+      for name in names.getElems do
+        if name.getId.toString.isPrefixOf hyp.userName.toString then
+          replaceMainGoal [← (← getMainGoal).clear hyp.fvarId]
+          continue
+
+
+
+section ForLean
+
+theorem BitVec.not_slt {w} (a b : BitVec w) : ¬ (a.slt b) ↔ (b.sle a) := by
```

Comment:
More to Arm/Bitvec.lean with a TODO to upstream?

Reply:
Done!